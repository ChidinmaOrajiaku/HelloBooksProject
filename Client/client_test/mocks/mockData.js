import { updatePassword } from "../../actions/updatePassword";

const mockData = {
  userDetails: {
    message: 'Welcome admin',
    username: 'admin',
    token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6ImRhZGR5Iiwicm9sZSI6InVzZXIiLCJpZCI6MiwiaWF0IjoxNTEzODc1ODE3LCJleHAiOjE1MTM5NjIyMTd9.oexUFhi08vhMU88LmZJvRDy9JFdThMpfb3Qz9DWWzgk'
  },
  userDetailsResponse: {
    username: 'daddy',
    role: 'user',
    id: 2,
    iat: 1513875817,
    exp: 1513962217
  },
  userCountResponse: {
    count: 7
  },
  borrowBookData: {
    bookId: 1,
    userId: 1,
  },
  borrowBookDataResponse: {
    returned: false,
    id: 24,
    userId: 1,
    bookId: 7,
    toReturnDate: '2018-01-14',
    updatedAt: '2017-12-21T21:16:08.442Z',
    createdAt: '2017-12-21T21:16:08.442Z',
    returnDate: null,
    categoryId: null
  },
  createBookData: {
    title: 'a',
    author: 'a',
    category: 'a',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1511133007/g2exfx6sgbymszspybts.jpg',
    review: 'Nice'
  },
  createBookErrorData: {
    author: 'a',
    category: 'a',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1511133007/g2exfx6sgbymszspybts.jpg',
    review: 'Nice'
  },
  createBookErrorResponse: 'An error occured',
  createBookResponse: {
    message: 'Succesfully added'
  },
  getABookData: {
    id: 7,
    title: 'Rage of Angels',
    author: 'Sidney Sheldon',
    category: 'Fiction',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1513096198/prbdr6eghnnrcewwpodf.jpg',
    review: 'Nice Book',
    createdAt: '2017-12-12T11:11:10.823Z',
    updatedAt: '2017-12-12T16:30:00.352Z',
    userId: null,
    categoryId: null
  },
  getAllBooksData: {
    id: 7,
    title: 'Rage of Angels',
    author: 'Sidney Sheldon',
    category: 'Fiction',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1513096198/prbdr6eghnnrcewwpodf.jpg',
    review: 'Nice Book',
    createdAt: '2017-12-12T11:11:10.823Z',
    updatedAt: '2017-12-12T16:30:00.352Z',
    userId: null,
    categoryId: null
  },
  getAllBorrowedBooksData: {
    id: 7,
    title: 'Rage of Angels',
    author: 'Sidney Sheldon',
    category: 'Fiction',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1513096198/prbdr6eghnnrcewwpodf.jpg',
    review: 'Nice Book',
    createdAt: '2017-12-12T11:11:10.823Z',
    updatedAt: '2017-12-12T16:30:00.352Z',
    userId: null,
    categoryId: null
  },
  getCategoryData: {
    category: 'Fiction'
  },
  getUserData: {
    firstname: 'daddy',
    lastname: 'daddy',
    username: 'daddy',
    password: 'daddy',
    email: 'daddy@gmail.com'
  },
  bookData: {
    title: 'a',
    author: 'b',
    category: 'a',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1511133007/g2exfx6sgbymszspybts.jpg',
    review: 'Nice'
  },
  returnBookData: {
    message: 'Successfully Returned'
  },
  userData: {
    verifyPassword: '1234',
    password: '4321'
  },
  errorMessage: 'An error occurred',
  image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1515403427/ena7dhy6xriztom6vuvo.jpg'
};

export default mockData;
