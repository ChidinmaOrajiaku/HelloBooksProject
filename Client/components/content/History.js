import React from 'react';
import axios from 'axios';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { createStore } from 'redux';
import { store } from '../../index';
import { getRequest } from '../../actions/booksAction';

class History extends React.Component {
 
  constructor (props) {
    super(props);
    this.state = {
      data: [],
    }
  }
  componentWillMount() {
    console.log(store)
      const user = store.getState()
      const userId = user.auth.user.id
      axios.get('/api/v1/users/' + userId + '/history').then((res) => {
        localStorage.getItem('jwtToken');
        this.setState({ data: res.data})
      });
   };
   
  render() {
  $(document).ready(function(){
    $('ul.tabs').tabs({
      swipeable : true,
      responsiveThreshold : 1920
    });
  });
  return (
    <div className="history">
    <table className="bordered highlight centered container">
    <thead>
      <tr>
          <th>Title</th>
          <th>Returned</th>
          <th>To Return Date</th>
          <th>Returned Date</th>
      </tr>
    </thead>

    <tbody>
    {
       this.state.data.map (
        history => 
      <tr>
        <td><div className = "row valign-wrapper"><img src={history.image}/> <p>{history.title} <br/> {history.author}</p></div></td>
        <td>{`${history.returned}`}</td>
        <td>{history.toReturnDate}</td>
        <td>{history.returnDate}</td>
      </tr>
       )
    }
    </tbody>
  </table>
  </div>
  );
};
}


export default connect(null,) (History);
