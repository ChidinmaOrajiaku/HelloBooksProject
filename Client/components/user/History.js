import React from 'react';
import axios from 'axios';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import NavigationBar from '../NavigationBar';
import { getUserBorrowed } from '../../actions/getUserBorrowedBooks';
import { yetToReturn } from '../../actions/yetToReturn';
import { returnBook } from '../../actions/returnBook';


/**
 * 
 * 
 * @class History
 * @extends {React.Component}
 */
class History extends React.Component {
  /**
  * Creates an instance of History.
  * @param {any} props 
  * @memberof History
  */
  constructor(props) {
    super(props);
    this.state = {
      userBorrowed: [],
      yetToReturn: [],
      bookStatus: '',
      loading: true,
      booksId: '',
      filterYetData: []
    };
    this.handleReturn = this.handleReturn.bind(this);
  }
  /**
 * 
 * @returns {event} handles change
 * @param {any} event 
 * @memberof History
 */
  handleChange(event) {
    this.setState({ [event.target.id]: event.target.value });
  }
  /**
 * 
 * @returns {event} handles return
 * @param {any} event 
 * @memberof History
 */
  handleReturn(event) {
    event.preventDefault();
    localStorage.setItem('booksId', event.target.value);
    this.setState({
      booksId: localStorage.getItem('booksId'),
      bookIndex: event.target.dataset.index
    }),
    setTimeout(() => {
      this.props.returnBook(this.props.usersId, this.state).then(() => {
        this.state.filterYetData.splice(this.state.bookIndex, 1);
        this.setState({
          yetToReturn: this.state.filterYetData
        });
        Materialize.toast('Succesfully Returned', 2000, 'teal rounded');
      }
      ).catch(() => {
        Materialize.toast('I dunno', 500, 'red rounded');
      });
    });
  }
  /**
 * 
 * @returns {data} data
 * @param {any} event 
 * @memberof History
 */
  componentDidMount() {
    this.props.getUserBorrowed(this.props.usersId);
    this.props.yetToReturn(this.props.usersId);
    $('select').material_select();
    $('select').change(event => this.handleChange(event));
  }

  /**
   * 
   * @returns {nextProps} next props
   * @param {any} nextProps 
   * @memberof Library
   */
  componentWillReceiveProps(nextProps) {
    this.setState({
      loading: false,
      userBorrowed: nextProps.getUserBorrowedData,
      yetToReturn: nextProps.yetToReturnData,
      filterYetData: nextProps.yetToReturnData,
    });
  }

  /**
     * 
     * 
     * @returns {ReactElement} Markup 
     * @memberof History
     */
  render() {
    const borrowedBooks = this.state.loading ? <div><p>Loading...</p></div> :
      <table className="bordered centered responsive-table">
        <thead>
          <tr className="white-text">
            <th> Image </th>
            <th>Title</th>
            <th>Author</th>
            <th>Category</th>
            <th> Returned </th>
            <th> Return Date </th>
            <th> Date Returned </th>
          </tr>
        </thead>
        <tbody>
          { Object.keys(this.state.userBorrowed).map(key =>
            <tr key={key}>
              <td> <img src = {this.state.userBorrowed[key].Book.image }/> </td>
              <td>{ this.state.userBorrowed[key].Book.title }</td>
              <td>{ this.state.userBorrowed[key].Book.author }</td>
              <td>{ this.state.userBorrowed[key].Book.category }</td>
              <td>{ this.state.userBorrowed[key].returned === false ? 'Not Returned' : 'Returned' }</td>
              <td>{ this.state.userBorrowed[key].toReturnDate === null ? 'N/A' : this.state.userBorrowed[key].toReturnDate }</td>
              <td>{ this.state.userBorrowed[key].returnDate === null ? 'N/A' : this.state.userBorrowed[key].returnDate }</td>
            </tr>
          )}
        </tbody>
      </table>;
    const yetToReturnBooks = this.state.loading ? <div><p>Loading...</p></div> :
      <table className="bordered centered responsive-table">
        <thead>
          <tr className="white-text">
            <th> Image </th>
            <th>Title</th>
            <th>Author</th>
            <th>Category</th>
            <th> Return </th>
          </tr>
        </thead>
        <tbody>
          { Object.keys(this.state.yetToReturn).map(key =>
            <tr key={key}>
              <td> <img src = {this.state.yetToReturn[key].Book.image }/> </td>
              <td>{ this.state.yetToReturn[key].Book.title }</td>
              <td>{ this.state.yetToReturn[key].Book.author }</td>
              <td>{ this.state.yetToReturn[key].Book.category }</td>
              <td><button value={this.state.yetToReturn[key].booksId} onClick={this.handleReturn} data-index= {key} className="waves-effect waves-light btn">Return</button></td>
            </tr>
          )}
        </tbody>
      </table>;
    return (
      <div className="history row container-fluid">
        <div className=""> <NavigationBar /> </div>
        <h4 className="col m8 offset-m3"> USER BOOK HISTORY </h4>
        <div className="row col m8 offset-m3">
          <div className="input-field col s4 status">
            <select className="teal-text" id= "bookStatus"value="1" onChange={this.handleChange}>
              <option value="" defaultValue>Choose your option</option>
              <option value="2">Borrowed Books</option>
              <option value="3">Pending Returns</option>
            </select>
            <label className="black-text sort">Sort table</label>
          </div>
          <div className="col s12 ">
            <div className="">
              <div className="card-content teal-text">
                {this.state.bookStatus === '3' ? yetToReturnBooks : borrowedBooks }
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => (
  {
    usersId: state.auth.user.id,
    getUserBorrowedData: state.userBorrowedBooks[0].response,
    yetToReturnData: state.yetToReturn[0].response,
  }
);

export default connect(mapStateToProps, { yetToReturn, getUserBorrowed, returnBook })(History);
