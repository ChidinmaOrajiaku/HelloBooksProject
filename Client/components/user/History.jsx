import React from 'react';
import { connect } from 'react-redux';
import Footer from '../Footer';
import NavigationBar from '../NavigationBar';
import { getUserBorrowed } from '../../actions/getUserBorrowedBooks';
import { yetToReturn } from '../../actions/yetToReturn';
import { returnBook } from '../../actions/returnBook';


/**
 *
 *
 * @class History
 * @extends {React.Component}
 */
export class History extends React.Component {
  /**
  * Creates an instance of History.
  * @constructor
  * @param {object} props
  * @memberof History
  */
  constructor(props) {
    super(props);
    this.state = {
      userBorrowed: [],
      yetToReturn: [],
      bookStatus: '',
      loading: true,
      bookId: '',
      filterYetData: []
    };
    this.handleReturn = this.handleReturn.bind(this);
  }

  /**
 * Mounts action on component
 * @returns {object} response object
 * @memberof History
 */
  componentDidMount() {
    this.props.getUserBorrowed(this.props.usersId);
    this.props.yetToReturn(this.props.usersId);
    $('select').material_select();
    $('select').change(event => this.handleChange(event));
  }

  /**
   * Receive nextprops and sets state
   * @param {object} nextProps
   * @returns {object} response object
   * @memberof History
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.getUserBorrowedData.message !== 'No books in the library') {
      this.setState({
        loading: false,
        userBorrowed: nextProps.getUserBorrowedData,
      });
    }
    if (nextProps.yetToReturnData !== 'No books in the library') {
      this.setState({
        loading: false,
        yetToReturn: nextProps.yetToReturnData,
        filterYetData: nextProps.yetToReturnData,
      });
    }
  }

  /**
 * Handles change and sets the state to the targeted event value
 * @returns {object} response object
 * @param {object} event
 * @memberof History
 */
  handleChange(event) {
    this.setState({ [event.target.id]: event.target.value });
  }

  /**
 * Handles books return, sets the state of the book Id and book index,
 splices book returned from list and sets the state of yet-to-return-books
 * @returns {object} resonse object
 * @param {object} event
 * @memberof History
 */
  handleReturn(event) {
    event.preventDefault();
    localStorage.setItem('bookId', event.target.value);
    this.setState({
      bookId: localStorage.getItem('bookId'),
      bookIndex: event.target.dataset.index
    }),
    setTimeout(() => {
      this.props.returnBook(this.props.usersId, this.state).then(() => {
        if (this.props.returnBookData.hasReturned === true) {
          this.state.filterYetData.splice(this.state.bookIndex, 1);
          this.setState({
            yetToReturn: this.state.filterYetData
          });
          Materialize.toast('Succesfully Returned', 2000, 'teal rounded');
        } else {
          Materialize.toast('Not Returned', 2000, 'red rounded');
        }
      });
    }, 1000);
  }

  /**
     * Renders component
     * @returns {object} ReactElementMarkup
     * @memberof History
     */
  render() {
    const borrowedBooks = this.state.loading ? <div><p>Loading...</p></div> :
      <table className="bordered centered responsive-table">
        <thead>
          <tr className="white-text">
            <th> Image </th>
            <th>Title</th>
            <th>Author</th>
            <th>Category</th>
            <th> Returned </th>
            <th> Return Date </th>
            <th> Date Returned </th>
          </tr>
        </thead>
        <tbody>
          { Object.keys(this.state.userBorrowed).map(key =>
            <tr key={key}>
              <td> <img src = {this.state.userBorrowed[key].Book.image }/> </td>
              <td>{ this.state.userBorrowed[key].Book.title }</td>
              <td>{ this.state.userBorrowed[key].Book.author }</td>
              <td>{ this.state.userBorrowed[key].Book.category }</td>
              <td>{ this.state.userBorrowed[key].returned === false ?
                'Not Returned' : 'Returned' }</td>
              <td>{ this.state.userBorrowed[key].toReturnDate === null ?
                'N/A' : this.state.userBorrowed[key].toReturnDate }
              </td>
              <td>{ this.state.userBorrowed[key].returnDate === null ?
                'N/A' : this.state.userBorrowed[key].returnDate }
              </td>
            </tr>)}
        </tbody>
      </table>;
    const yetToReturnBooks = this.state.yetToReturn.message === 'No books in the library' ? <div><p>No pending returns</p></div> :
      <table className="bordered centered responsive-table">
        <thead>
          <tr className="white-text">
            <th> Image </th>
            <th>Title</th>
            <th>Author</th>
            <th>Category</th>
            <th> Return </th>
          </tr>
        </thead>
        <tbody>
          { Object.keys(this.state.yetToReturn).map(key =>
            <tr key={key}>
              <td> <img src = {this.state.yetToReturn[key].Book.image }/> </td>
              <td>{ this.state.yetToReturn[key].Book.title }</td>
              <td>{ this.state.yetToReturn[key].Book.author }</td>
              <td>{ this.state.yetToReturn[key].Book.category }</td>
              <td><button value={this.state.yetToReturn[key].bookId}
                id="returnAction"
                onClick={this.handleReturn} data-index= {key}
                className="waves-effect waves-light btn">Return</button>
              </td>
            </tr>)}
        </tbody>
      </table>;

    return (
      <div className="history row container-fluid">
        <div className=""> <NavigationBar /> </div>
        <h4 className="col s12 m8 offset-m3 white-text"> USER BOOK HISTORY </h4>
        <div className="row col m8 offset-m3">
          <div className="input-field col s4 status">
            <select className="white-text" id= "bookStatus"value="1"
              onChange={this.handleChange}>
              <option value="" defaultValue>Choose your option</option>
              <option value="2">Borrowed Books</option>
              <option value="3">Pending Returns</option>
            </select>
            <label className="white-text sort">Sort table</label>
          </div>
          <div className="col s12 ">
            <div className="card">
              <div className="card-content teal-text">
                {this.state.bookStatus === '3' ?
                  yetToReturnBooks : borrowedBooks }
              </div>
            </div>
          </div>
          <div> <Footer /></div>
        </div>
      </div>
    );
  }
}

export const mapStateToProps = state => (
  {
    usersId: state.auth.user.id,
    getUserBorrowedData: state.userBorrowedBooks[0].response,
    yetToReturnData: state.yetToReturn[0].response,
    returnBookData: state.returnBook[0]
  }
);

export default connect(mapStateToProps, {
  yetToReturn,
  getUserBorrowed,
  returnBook
})(History);
