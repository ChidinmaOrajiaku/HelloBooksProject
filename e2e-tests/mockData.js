module.exports = {
  file: {
    lastModified: 1508779018000,
    lastModifiedDate: 'Mon Oct 23 2017 18:16:58 GMT+0100 (WAT)',
    name: '3 card game.jpg',
    size: 642971,
    type: 'image/jpeg',
    webkitRelativePath: ''
  },
};

