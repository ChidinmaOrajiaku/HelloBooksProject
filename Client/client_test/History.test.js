import React from 'react';
import $ from 'jquery';
import toJson from 'enzyme-to-json';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import mockData from './mocks/mockData';
import { History } from '../components/user/History.jsx';

configure({ adapter: new Adapter() });

describe('<History />', () => {
  const props = {
    getUserBorrowed: jest.fn(),
    yetToReturn: jest.fn(),
  };
  let mountedComponent;
  const navigationBarItem = () => {
    if (!mountedComponent) {
      mountedComponent = shallow(<History {...props}/>);
    }
    return mountedComponent;
  };

  it('should call handleReturn method', () => {
    const spy = jest.spyOn(navigationBarItem().instance(), 'handleReturn');
    const event = {
      preventDefault: jest.fn(),
      target: {
        value: 'abd',
        dataset: {
          index: 1
        }
      }
    };
    navigationBarItem().instance().handleReturn(event);
    expect(spy).toHaveBeenCalled();
  });

  it('should call handleChange method', () => {
    const spy = jest.spyOn(navigationBarItem().instance(), 'handleChange');
    const event = {
      preventDefault: jest.fn(),
      target: {
        id: 'review',
        value: 'chideberecom'
      }
    };
    navigationBarItem().instance().handleChange(event);
    expect(spy).toHaveBeenCalled();
  });

  it('should match snapshot test', () => {
    const component = navigationBarItem();
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
