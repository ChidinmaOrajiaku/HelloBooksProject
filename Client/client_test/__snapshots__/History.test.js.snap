// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<History /> should match snapshot test 1`] = `
<div
  className="history row container-fluid"
>
  <div
    className=""
  >
     
    <Connect(withRouter(NavigationBar)) />
     
  </div>
  <h4
    className="col m8 offset-m3 white-text"
  >
     USER BOOK HISTORY 
  </h4>
  <div
    className="row col m8 offset-m3"
  >
    <div
      className="input-field col s4 status"
    >
      <select
        className="white-text"
        id="bookStatus"
        onChange={[Function]}
        value="1"
      >
        <option
          defaultValue={true}
          value=""
        >
          Choose your option
        </option>
        <option
          value="2"
        >
          Borrowed Books
        </option>
        <option
          value="3"
        >
          Pending Returns
        </option>
      </select>
      <label
        className="white-text sort"
      >
        Sort table
      </label>
    </div>
    <div
      className="col s12 "
    >
      <div
        className="card"
      >
        <div
          className="card-content teal-text"
        >
          <div>
            <p>
              Loading...
            </p>
          </div>
        </div>
      </div>
    </div>
    <div>
       
      <Component />
    </div>
  </div>
</div>
`;
